java-checks/src/main/java/org/sonar/java/checks/UnreachableCatchCheck.java

```diff
@@ -0,0 +1,164 @@
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.UnionTypeTree;
+
+@Rule(key = "S4970")
+public class UnreachableCatchCheck extends IssuableSubscriptionVisitor {
+
+  private Map<Type, SyntaxToken> typeToCatchToken = new HashMap<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.TRY_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    TryStatementTree tryStatementTree = (TryStatementTree) tree;
+    typeToCatchToken.clear();
+    Multimap<Type, Type> derivedToBase = getDerivedTypesCaughtBeforeBaseType(tryStatementTree.catches());
+
+    if (derivedToBase.isEmpty()) {
+      return;
+    }
+
+    MethodThrownTypeCollector collector = new MethodThrownTypeCollector();
+    tryStatementTree.block().accept(collector);
+    List<Type> thrownTypes = collector.thrownTypes;
+
+    derivedToBase.entries().forEach(entry -> {
+      Type derivedType = entry.getKey();
+      Type baseType = entry.getValue();
+
+      // Catching a derived type before the base type is fine if the body of the try throws an exception which is a subtype of the base type,
+      // but not of the derived type. We have to make sure that we are not in this situation before reporting an issue.
+      if (isHidden(baseType, derivedType, thrownTypes)) {
+        reportIssue(typeToCatchToken.get(baseType),
+          "Remove this catch block because it is unreachable as hidden by previous catch blocks.",
+          Collections.singletonList(new JavaFileScannerContext.Location("Already catch the exception", typeToCatchToken.get(derivedType))),
+          null);
+      }
+    });
+  }
+
+  private Multimap<Type, Type> getDerivedTypesCaughtBeforeBaseType(List<CatchTree> catches) {
+    Multimap<Type, Type> derivedBeforeBase = HashMultimap.create();
+
+    List<Type> catchTypes = catches.stream()
+      .flatMap(c -> {
+        List<Type> types = new ArrayList<>();
+        collectTypesFromTypeTree(c.parameter().type(), types, c.catchKeyword());
+        return types.stream();
+      })
+      .filter(UnreachableCatchCheck::isChecked)
+      .collect(Collectors.toList());
+
+    for (int i = 0; i < catchTypes.size() - 1; i++) {
```

Comment:
In fact, the current implementation reports duplicates! I thought I had a test for it, but the next line is `Type topType = catchTypes.get(0);` instead of `catchTypes.get(i);` 😕 

I changed the implementation to inverse the key and the value, the map will now look like:
```
D: A,B,C
C: B,A
B: A
```

We still have the map though, we could slightly change the logic to avoid it, but we will have to add extra logic for the same result, I believe the current status is good enough.